src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSenderPool.cs

```diff
@@ -0,0 +1,55 @@
+using System;
+using System.Collections.Concurrent;
+using System.IO.Pipelines;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
+{
+    internal class SocketSenderPool : IDisposable
+    {
+        private const int MaxQueueSize = 1024; // REVIEW: Is this good enough?
+
+        private readonly ConcurrentQueue<SocketSender> _queue = new();
+        private readonly PipeScheduler _scheduler;
+        private bool _disposed;
+
+        public SocketSenderPool(PipeScheduler scheduler)
+        {
+            _scheduler = scheduler;
+        }
+
+        public SocketSender Rent()
+        {
+            if (_queue.TryDequeue(out var sender))
+            {
+                return sender;
+            }
+            return new SocketSender(_scheduler);
+        }
+
+        public void Return(SocketSender sender)
+        {
+            if (_disposed || _queue.Count > MaxQueueSize)
+            {
+                sender.Dispose();
+                return;
+            }
+
+            sender.Reset();
+
+            _queue.Enqueue(sender);
+        }
+
+        public void Dispose()
+        {
+            if (!_disposed)
```

Comment:
Is thread safety an issue when disposing? What happens if there is a sender that hasn't been returned to the queue?

Reply:
That's not a concern. It'll probably leak. We have that same issue with the memory pool.

Reply:
To clarify, in the face of ungraceful shutdown, bad things can happen with both the memory pool and now this pool.